ffmuc-mesh-vpn-wireguard-vxlan/files/lib/gluon/gluon-mesh-wireguard-vxlan/checkuplink

```diff
@@ -65,33 +83,78 @@ if [ "$(uci get wireguard.mesh_vpn.enabled)" == "true" ] || [ "$(uci get wiregua
 		NTP_SERVER=$(uci get system.ntp.server)
 		gluon-wan /usr/sbin/ntpd -n -N -S /usr/sbin/ntpd-hotplug -p $NTP_SERVER -q
 
-		# Get the number of configured peers and randomly select one
-		NUMBER_OF_PEERS=$(uci -q show wireguard | egrep -ce peer_[0-9]+.endpoint)
-		PEER="$(awk -v min=1 -v max=$NUMBER_OF_PEERS 'BEGIN{srand(); print int(min+rand()*(max-min+1))}')"
-		PEER_PUBLICKEY="$(uci get wireguard.peer_$PEER.publickey)"
-
-		logger -t checkuplink "Selected peer $PEER"
-
-		endpoint="$(check_address_family "$PEER_PUBLICKEY" "$(uci get wireguard.peer_$PEER.endpoint)")"
-
-		logger -t checkuplink "Connecting to $endpoint"
-
 		# Delete Interfaces
 		ip link set nomaster dev mesh-vpn &> /dev/null
 		ip link delete dev mesh-vpn &> /dev/null
 		ip link del $MESH_VPN_IFACE &> /dev/null
 		PUBLICKEY=$(uci get wireguard.mesh_vpn.privatekey | wg pubkey)
 		SEGMENT=$(uci get gluon.core.domain)
 
-		# Push public key to broker, test for https and use if supported
+		# Push public key to broker and receive gateway data, test for https and use if supported
 		wget -q https://[::1]
 		if [ $? -eq 1 ]; then
 			PROTO=http
 		else
 			PROTO=https
 		fi
-		gluon-wan wget -q  -O- --post-data='{"domain": "'"$SEGMENT"'","public_key": "'"$PUBLICKEY"'"}' $PROTO://$(uci get wireguard.mesh_vpn.broker)
-		
+
+		if [ "$(uci get wireguard.mesh_vpn.loadbalancing)" == "true" ] || [ "$(uci get wireguard.mesh_vpn.loadbalancing)" == "1" ]; then
+			# Use /api/v2, get gateway peer details from broker response
+			WGKEX_BROKER="$PROTO://$(uci get wireguard.mesh_vpn.broker)/api/v2/wg/key/exchange"
+			logger -t checkuplink "Contacting wgkex broker $WGKEX_BROKER"
+			WGKEX_DATA=$(gluon-wan wget -q -O- --post-data='{"domain": "'"$SEGMENT"'","public_key": "'"$PUBLICKEY"'"}' "$WGKEX_BROKER")
+			if [ $? -eq 1 ]; then
+				logger -t checkuplink "Contacting wgkex broker failed, response: $WGKEX_DATA"
+				exit 1
+			fi
+
+			logger -t checkuplink "Got data from wgkex broker: $WGKEX_DATA"
+
+			data=$(lua /lib/gluon/gluon-mesh-wireguard-vxlan/parse-wgkex-response.lua "$WGKEX_DATA")
+			if [ $? -eq 1 ]; then
+				logger -t checkuplink "Parsing wgkex broker data failed"
+				exit 1
+			fi
+
+			logger -t checkuplink "Parsed wgkex broker data"
+
+			PEER_ADDRESS=$(echo "$data" | sed -n 1p)
+			PEER_PORT=$(echo "$data" | sed -n 2p)
+			PEER_PUBLICKEY=$(echo "$data" | sed -n 3p)
+			PEER_LINKADDRESS=$(echo "$data" | sed -n 4p)
+
+			PEER_ADDRESS=$(resolve_host "$PEER_ADDRESS")
+			PEER_ENDPOINT=$(combine_ip_port "$PEER_ADDRESS" "$PEER_PORT")
+
+		else
+			# Use /api/v2, get gateway peer details from config
+			WGKEX_BROKER="$PROTO://$(uci get wireguard.mesh_vpn.broker)/api/v1/wg/key/exchange"
+			logger -t checkuplink "Contacting wgkex broker $WGKEX_BROKER"
+			gluon-wan wget -q -O- --post-data='{"domain": "'"$SEGMENT"'","public_key": "'"$PUBLICKEY"'"}' "$WGKEX_BROKER"
+			if [ $? -eq 1 ]; then
+				logger -t checkuplink "Contacting wgkex broker failed"
+				exit 1
+			fi
```

Comment:
Can we deduplicate this code here with the one above?

Reply:
If we want to deduplicate, I'd do it using a function, as not to break up the the if-else blocks, because right now the logic is cleanly separated into two code paths for the different API versions.
Not sure if these few lines, which are mostly error handling, are worth it?

Reply:
We can also do this later if you want.